### ISSUE TITLE
Alleviate memory usage of ghidra, and PdbUniversal component

### ISSUE THREAD

**Issue by homes410** [NationalSecurityAgency/ghidra#3497]
**Describe the bug**
[0002-211013-objectcache.patch-2.txt](https://github.com/NationalSecurityAgency/ghidra/files/7335163/0002-211013-objectcache.patch-2.txt)
1.
Despite of resolution of issue #3141, Memory usage of ghidra grows so fast as you work on large file.
I found that *removing a strong reference to DBObjectCache from DatabaseObject improved overall response time of ghidra*. 
(just calling clear() on keyedsoftreference object when the object removed from hard cache whether it is removed from manager or not would also work with expense of more io operation. but that is not done in this patch.)
after this patch, memory usage rarely exceeds javaheap size with default setting of SoftRefLRUPolicyMSPerMB.

[0001-211012-pdbapplicator.patch.txt](https://github.com/NationalSecurityAgency/ghidra/files/7333844/0001-211012-pdbapplicator.patch.txt)
2. PdbUniversal component keep symbol reference in primarySymbolByAddress. in this patch it will hold id field of symbol instead of object itself.

Above patches can be applied to master branch of ghidra without a conflict.
Best regards.
**Environment (please complete the following information):**
 - Ghidra Version: 9.2.4

**Additional context**
[0001-211012-pdbapplicator.patch.txt](https://github.com/NationalSecurityAgency/ghidra/files/7333844/0001-211012-pdbapplicator.patch.txt)
[0002-211013-objectcache.patch-2.txt](https://github.com/NationalSecurityAgency/ghidra/files/7335163/0002-211013-objectcache.patch-2.txt)

**Comment by homes410** [comment]
> Why you saw memory improvements and I didn't is unclear to me. It could be what I stated earlier or, perhaps, because maybe your improvements were measured with both patches in play. Do you have profiling numbers specific to just the PDB change?

No I don't have any of these. because of those effects caused from two changes coincides each other.

**Comment by homes410** [comment]
> We would like you to try just applying your changes to the PDBApplicator without the db changes to see what behavior that produces. We think that the db changes may not be necessary.

below are time taken to analyze msedge.dll(>170MB) in ms edge browser with pdb symbol(320MB) with the patches applied.
I began to work on this patch after I see the analysis didn't finish in two days.

2021-10-13 21:55:10 INFO  ghidra.app.plugin.core.analysis.AutoAnalysisManager -----------------------------------------------------
    PDB Universal - One Time               69442.840 secs
    Scalar Operand References               2893.864 secs
-----------------------------------------------------
     Total Time   72336 secs
-----------------------------------------------------
  
2021-10-13 21:55:13 DEBUG ghidra.framework.plugintool.mgr.ToolTaskManager   task finish (72351.512 secs)  
2021-10-13 21:55:14 DEBUG ghidra.framework.plugintool.mgr.ToolTaskManager   task complete (72352.461 secs)  

without the db patch. , it would take more time because of gc overhead in jvm. -  no changes in memory usage of ghidra. but gc will spend more time.

**Comment by dragonmacher** [comment]
Are you able to run Ghidra from the master branch?   There are changes in master that should address the huge amount of time that analysis is taking.   We would like to know if the master branch changes fix all of the issues (time and memory) that you are seeing.

**Comment by ghizard** [comment]
Thank you for having supplied this patch.  I had investigated it and liked what you had supplied and had hoped to put a different fix in for this sooner.  Note, however, that there have been a number of recent fixes in 11.0 with regard to improving memory PDB performance which much outweigh this issue.

I've recently begun work in revisiting the logic for selecting the primary symbol at an address and while doing this, I revisited the performance of requesting the primary symbol from the symbol manager vs. my using a map to track it, and the underlying symbol manager methods are now much more performant.  Thus, I'm able to eliminate the map used in the PDB altogether.
